pkg/innerring/internal/blockchain/blockchain.go

```diff
@@ -0,0 +1,383 @@
+// Options tunes Blockchain's default behavior. All fields are optional: zero
+// (omitted) values don't change the behavior. Launched Blockchain should be
+// stopped.
+type Options struct {
+	// Writer of the Blockchain's logs and internal errors. By default, Blockchain
+	// doesn't write logs.
+	Logger *zap.Logger
+
+	// Network addresses to listen Neo P2P in 'host:port' format. By default, Neo
+	// P2P is not served.
+	P2pEndpoints []string
+
+	// List of nodes' addresses to communicate with over Neo P2P protocol in
+	// 'host:port' format. By default, node runs as standalone.
+	SeedNodes []string
+
+	// Network addresses to listen Neo RPC in 'host:port' format. By default, Neo
+	// RPC is not served.
+	RPCEndpoints []string
+
+	// Disable Notary service.
+	NotaryDisabled bool
+}
+
+// New returns new Blockchain configured by specified Config and Options.
+// Resulting Blockchain is ready to run.
+func New(cfg Config, opts Options) (*Blockchain, error) {
+	switch {
+	case cfg.Wallet.Path == "":
+		panic("missing wallet path")
+	case cfg.ErrorListener == nil:
+		panic("missing error channel")
+	case cfg.NetworkMagic == 0:
+		// actually zero magic is valid, but almost definitely forgotten
+		panic("missing network magic")
+	case len(cfg.Committee) == 0:
+		panic("empty committee")
+	}
+
+	if opts.Logger == nil {
+		opts.Logger = zap.NewNop()
+	}
+
+	nodeAcc, err := utilConfig.LoadAccount(cfg.Wallet.Path, "", cfg.Wallet.Password)
+	if err != nil {
+		return nil, fmt.Errorf("read node account: %w", err)
+	}
+
+	standByCommittee := make([]string, len(cfg.Committee))
+	for i := range cfg.Committee {
+		standByCommittee[i] = hex.EncodeToString(cfg.Committee[i].Bytes())
+	}
+
+	var cfgBase config.Config
+
+	cfgBaseProto := &cfgBase.ProtocolConfiguration
+	cfgBaseProto.Magic = cfg.NetworkMagic
+	cfgBaseProto.StandbyCommittee = standByCommittee
+	cfgBaseProto.ValidatorsCount = len(standByCommittee)
+	cfgBaseProto.TimePerBlock = time.Second
+	cfgBaseProto.SeedList = opts.SeedNodes
+	cfgBaseProto.VerifyTransactions = true
+	cfgBaseProto.P2PSigExtensions = !opts.NotaryDisabled
+
+	cfgBaseApp := &cfgBase.ApplicationConfiguration
+	cfgBaseApp.Relay = true
+	cfgBaseApp.UnlockWallet = cfg.Wallet
+	cfgBaseApp.Consensus.Enabled = true
+	cfgBaseApp.Consensus.UnlockWallet = cfg.Wallet
+	cfgBaseApp.P2PNotary.Enabled = !opts.NotaryDisabled
+	cfgBaseApp.P2PNotary.UnlockWallet = cfg.Wallet
+	cfgBaseApp.StateRoot.Enabled = true
+	cfgBaseApp.StateRoot.UnlockWallet = cfg.Wallet
+	cfgBaseApp.RPC.Enabled = true
+	cfgBaseApp.RPC.Addresses = opts.RPCEndpoints
+	cfgBaseApp.RPC.StartWhenSynchronized = true
+	cfgBaseApp.RPC.MaxGasInvoke = fixedn.Fixed8FromInt64(100)
+	cfgBaseApp.P2P.Addresses = opts.P2pEndpoints
+	cfgBaseApp.P2P.DialTimeout = 3 * time.Second
+	cfgBaseApp.P2P.ProtoTickInterval = 2 * time.Second
+	cfgBaseApp.P2P.PingInterval = 30 * time.Second
+	cfgBaseApp.P2P.PingTimeout = 90 * time.Second
+	cfgBaseApp.P2P.MaxPeers = 100
+	cfgBaseApp.P2P.AttemptConnPeers = 20
```

Comment:
These P2P params are for public networks, but they may be different for single node or for special setups. MinPeers is also missing, but it's very important for `StartWhenSynchronized` to work correctly.